sql/src/main/java/com/rilldata/calcite/CalciteToolbox.java

```diff
@@ -131,6 +165,61 @@ public SqlValidator getValidator(PlannerImpl planner) throws NoSuchFieldExceptio
     return (SqlValidator) validatorField.get(planner);
   }
 
+  static class Statement
+  {
+    public List<Statement> dependencies;
+    SqlNode node;
+    String ddl;
+    String name;
+    boolean changed;
+
+    public String getName()
+    {
+      return name;
+    }
+
+    public String getType()
+    {
+      if (node instanceof SqlCreateTable) {
+        return "TABLE";
+      } else if (node instanceof SqlCreateView)
+        return "VIEW";
+      return null;
+    }
+  }
+
+  public static List<MigrationStep> inferMigrations(String newSql, String schema, SqlDialect sqlDialect)
+      throws SqlParseException, JsonProcessingException
+  {
+    SqlNode node = parseStmts(newSql);
+
+    Map<String, Statement> existing = toStatementsMap(schema, sqlDialect);
+    createGraph(existing);
+    Map<String, Statement> ast = toStatementsMap(node, sqlDialect);
+    Set<String> seen = new HashSet<>();
+    List<MigrationStep> steps = existing.values().stream().filter(s -> s.getType() != null).map(s -> new MigrationStep(s.getName(), s.getType())).collect(
+        Collectors.toList());
+    steps.addAll(ast.values().stream().filter(s -> s.ddl != null).map(s -> new MigrationStep(s.ddl)).collect(
+        Collectors.toList()));
+
+    for (Statement create : ast.values()) {
+      if (existing.keySet().contains(create.getName())) {
+        Statement migrateFrom = existing.get(create.name);
+        create.changed = !create.ddl.equals(migrateFrom.ddl);
+        if (create.changed) {
+// todo         markDependentsChange(migrateFro);
+        }
+      }
```

Comment:
what happens to statements that are present in existing but not in new ast ?

Reply:
they will be converted to drop statements, see com.rilldata.calcite.CalciteToolboxTest#testMigrations